Add Kubernetes cluster validation

girikuncoro opened the pull request:
**What this PR does / why we need it**:
This PR adds Kubernetes cluster validation as part of clusterctl validate cluster

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
Issue #168

girikuncoro commented:
can you do quick check if I'm going to right direction? if it's fine, I will continue to write test for this. also can you guide me on how to run test? I was trying to run `go test` from `cmd/clusterctl/validation` but failed (it says something like cannot start control plane, exceed retry)

/hold
/cc @wangzhen127

girikuncoro commented:
/assign @wangzhen127

vincepri commented on cmd/clusterctl/cmd/validate_cluster.go:
```diff
@@ -76,7 +77,9 @@ func RunValidateCluster() error {
 	if err = validation.ValidateClusterAPIObjects(os.Stdout, c, vco.KubeconfigOverrides.Context.Cluster, vco.KubeconfigOverrides.Context.Namespace); err != nil {
 		return err
 	}
+	if err = validation.ValidateKubeClusterObjects(os.Stdout, c, vco.KubeconfigOverrides.Context.Cluster); err != nil {
```
The inline assignment should be `err :=` to avoid overwriting the `err` outside of the `if` scope.

vincepri replied:
Noticed the aline above shows the same pattern, good opportunity to fix it there too :)

vincepri commented on cmd/clusterctl/validation/validate_kube_cluster_objects.go:
```diff
@@ -0,0 +1,45 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func ValidateKubeClusterObjects(w io.Writer, c client.Client, clusterName string) error {
```
We should add unit tests for this function and the one below.

enxebre commented on cmd/clusterctl/validation/validate_kube_cluster_objects.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+func ValidateKubeClusterObjects(w io.Writer, c client.Client, clusterName string) error {
+	fmt.Fprintf(w, "Validating Kube Cluster objects\n")
```
can we use klog for consistency?

girikuncoro commented:
@wangzhen127 I have addressed all comments except the unit test. Currently debugging this first before writing my own tests
```
$ go test
2019/01/21 10:12:38 failed to start the controlplane. retried 5 times
exit status 1
FAIL    sigs.k8s.io/cluster-api/cmd/clusterctl/validation       0.061s
```

girikuncoro commented:
@wangzhen127 I have refactored a bit and added test, can you review again? thanks!

vincepri commented:
@girikuncoro yes please, running `make gazelle` should do the trick.

girikuncoro commented:
@vincepri @wangzhen127 bazel file updated, please kindly review, thanks!

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+func ValidatePods(w io.Writer, c client.Client) error {
+	fmt.Fprintf(w, "Validating pods\n")
+
+	pods, err := getPods(c, "kube-system")
```
Can the namespace be passed in instead of hard code here?

girikuncoro replied:
I have passed `kube-system` from `cmd/validate_cluster.go` instead if this is fine https://github.com/kubernetes-sigs/cluster-api/pull/679/commits/c77856e21219fa6c6a4969b810168786e6e3978b#diff-6334f0ae8a2b2bebd8b635352f2a9426R79

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+
+	if len(pods.Items) == 0 {
+		fmt.Fprintf(w, "FAIL\n")
+		fmt.Fprintf(w, "\tpods not exist.\n")
```
The error message can be enhanced, e.g., include in which namespace the pods do not exist.

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+		if pod.Status.Phase == corev1.PodPending ||
+			pod.Status.Phase == corev1.PodFailed ||
+			pod.Status.Phase == corev1.PodUnknown {
+
```
nit: empty line not needed

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+		}
+	}
+
+	if len(pods.Items) == 0 {
```
This check should be at the beginning of this function.

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+
+func validatePods(w io.Writer, pods *corev1.PodList) error {
+	for _, pod := range pods.Items {
+		fmt.Fprintf(w, "Checking pod %q in namespace %q...", pod.Name, pod.Namespace)
```
This is too verbose. If there are 100 pods in the namespace, there will be too many messages printed out. Just one message outside the for loop is enough.

wangzhen127 commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,81 @@
+			pod.Status.Phase == corev1.PodFailed ||
+			pod.Status.Phase == corev1.PodUnknown {
+
+			fmt.Fprintf(w, "FAIL\n")
```
If there are many failed pods, there will be too many "FAIL" messages. I suggest collecting all the errors and save in some data structure. And then print "FAIL" if there is some error at the end.

Same for containers below.

girikuncoro replied:
I don't think this will happen, since the function will return early when first pod failure happens. But if what you prefer is to collect all error / pod failures into some datastructure first by looping the whole pod list and check for error at the end, I'll do that since it seems cleaner

girikuncoro commented:
@wangzhen127 I've addressed all your comments, please check again

girikuncoro commented:
/test pull-cluster-api-test

girikuncoro commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,97 @@
+
+	if len(failures) != 0 {
+		fmt.Fprintf(w, "FAIL\n")
+		fmt.Fprintf(w, "\t[%v]: %s\n", failures[0].name, failures[0].message)
```
@wangzhen127 I've collected all pod failures, do you prefer to print just the first one like this or anything else?

wangzhen127 replied:
I prefer to print all failures, so that users can see all of them.

girikuncoro commented:
@wangzhen127 I've addressed the comments, just one question related to printing the pod failures here, do you prefer to just print the first failure or any other ideas? https://github.com/kubernetes-sigs/cluster-api/pull/679/files#r250440087

girikuncoro commented:
@wangzhen127 solved, please check again and let me know if you find anything, thanks!

wangzhen127 commented:
Thanks! LGTM from my side.
/lgtm

roberthbailey commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
nit: 2019

girikuncoro replied:
Fixed here: https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-53767fe464d92e82c3034f3769d689dfR2

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
nit: 2019

girikuncoro replied:
Fixed here: https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-a363c04a3cb096347d5c77d779ec333fR2

roberthbailey commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,99 @@
+	message string
+}
+
+func ValidatePods(w io.Writer, c client.Client, namespace string) error {
```
Shouldn't this function take a context argument? The other functions that it calls use `context.TODO()` but that's more typical for existing code where context has yet to be plumbed through. This is new code, and I don't see any reason it shouldn't pass down context from the start.

girikuncoro replied:
am I right doing it like this? https://github.com/kubernetes-sigs/cluster-api/pull/679/commits/0123f104fe86be502d65326ed971e1fb5133dcc8#diff-6334f0ae8a2b2bebd8b635352f2a9426R81

I was just trying to make it consistent with the other validation function `ValidateClusterAPIObjects`, is this also opportunity to fix passing context here? if so, I'll address in another PR once this one gets merged.
https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/validation/validate_cluster_api_objects.go#L56
https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/validation/validate_cluster_api_objects.go#L46
https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/validation/validate_cluster_api_objects.go#L61

roberthbailey replied:
Yes, that looks better. Fixing the others in a follow-up PR sounds great.

roberthbailey commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,99 @@
+func getPods(c client.Client, namespace string) (*corev1.PodList, error) {
+	pods := &corev1.PodList{}
+	if err := c.List(context.TODO(), client.InNamespace(namespace), pods); err != nil {
+		return nil, fmt.Errorf("Failed to get pods in namespace %q: %v", namespace, err)
```
nit: error strings should start with a lower case letter. 

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

girikuncoro replied:
Fixes:
https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-53767fe464d92e82c3034f3769d689dfR47

https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-53767fe464d92e82c3034f3769d689dfR56

https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-53767fe464d92e82c3034f3769d689dfR94

roberthbailey commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,99 @@
+	return pods, nil
+}
+
+func validatePods(w io.Writer, pods *corev1.PodList, namespace string) error {
```
This validates that all pods are running, but it doesn't check for the existence of specific pods that need to be there or for extraneous pods whose health doesn't affect the health of the cluster.

girikuncoro replied:
@roberthbailey can you elaborate more on how to approach this? 
* I think this is not straight forward to implement, since we need to take number of nodes/masters into account which will impact number of apiserver, kubeproxy, etc. to validate. Also, need to consider internal vs external etcd.

* How do we check for existence of specific pods which are required? Do we check the labels? Are label values standard across k8s versions?

* What are considered required pods? I was thinking 6 types: apiserver, controller-manager, scheduler,  etcd (if internal), kube-proxy, kubedns/coredns. CNI is hard to validate. Also, how do you want to output extraneous pods which are failing, output as warning?

roberthbailey replied:
I agree with your points, but it doesn't make the current code make more sense to me. 

To expand on your first point, depending on how the cluster is structured, it isn't always possible to see the control plane pods (e.g. if they are running as static pods and the kubelet on the master isn't registered into the cluster). So you may not always be able to check for the apiserver, scheduler, etc. Also, it's possible to create a fully conformant cluster without running kube-proxy on nodes (if you have a different mechanism for doing service forwarding). 

Using pod names or labels doesn't seem like it's going to be a reliable way to find the pods we care about. In fact, "the pods we care about" is a provider-specific notion and while there may be some commonality, I'm not sure how much we can count on.

girikuncoro replied:
I have added component validation outside of validatePods function if this is acceptable https://github.com/kubernetes-sigs/cluster-api/pull/679/commits/2117bc7df65b7ca366ef679ddb5ea4a6d7e654f3

roberthbailey commented on cmd/clusterctl/validation/validate_pods.go:
```diff
@@ -0,0 +1,99 @@
+	"io"
+	"strings"
+
+	"golang.org/x/net/context"
```
why not just import "context"?

girikuncoro replied:
Fixed here:
https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-53767fe464d92e82c3034f3769d689dfR20

https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-6334f0ae8a2b2bebd8b635352f2a9426R20

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+func getPodWithContainerSpec(podName, namespace, containerName, containerImage string) corev1.Pod {
```
nit: don't start the function name with "get"; this should be `podWithContainerSpec(...)`

girikuncoro replied:
Fixed here for `podWithStatus ` method:
https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-a363c04a3cb096347d5c77d779ec333fR27

I have deleted `podWithContainerSpec` method since it's not required anymore

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+func getPodWithStatus(podName, namespace string, podPhase corev1.PodPhase, containerReadyStatus bool) corev1.Pod {
```
same with this one.

girikuncoro replied:
Fixed here:
https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-a363c04a3cb096347d5c77d779ec333fR27

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+func TestGetPods(t *testing.T) {
```
Is this a useful test? It seems to be testing the functionality of the manager and it's ability to query k8s and not any actual logic in this package.

girikuncoro replied:
I have deleted this test case since it's not testing the validation logic

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func TestValidatePodsWithNoPod(t *testing.T) {
+	pods := &corev1.PodList{
```
nit: This can all be on one line.

girikuncoro replied:
Fixed here: https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-a363c04a3cb096347d5c77d779ec333fR46

wangzhen127 commented:
Hold was added by @girikuncoro. So I will leave it to him.

roberthbailey commented on cmd/clusterctl/validation/validate_pods_test.go:
```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+func TestValidatePodsWithOnePod(t *testing.T) {
```
This test doesn't cover any of the looping functionality (first pod error, second ok and vice versa) nor error aggregation.

girikuncoro replied:
Addressed this in new test case `TestValidatePodsWithNPods`

girikuncoro replied:
Here: https://github.com/kubernetes-sigs/cluster-api/pull/679/files#diff-a363c04a3cb096347d5c77d779ec333fR119

girikuncoro commented:
/hold cancel

girikuncoro commented:
> You resolved a bunch of my comments without actually fixing the code (or giving a reason). Please address all comments, either by updating the code or explaining why you disagree.

@roberthbailey I actually resolved your comments after fixing the code, but sorry if it confused you since I didn't write anything. I have put link to fixes on each of your comment that I addressed.

girikuncoro commented:
@roberthbailey I have added component validation as separate function if acceptable, and have put comments on each of your review, can you check again?

roberthbailey commented:
/lgtm
